kvclient: decompose CPut requests in the txnWriteBuffer

arulajmani opened the pull request:
This patch decomposes CPut requests, which are read-write in nature, into a read half and a write half. This is done by sending a (Exclusive) locking Get request to KV, followed by the evaluation of the CPut conditional on the client. If successful, the associated write is buffered on the client. Subsequently, at flush time, this buffered write will be converted to a Put request.

If the conditional evaluation isn't successful, an ConditionFailedError is returned back. The layers above should be oblivious to this transformation in the txnWriteBuffer; so the ConditionFailedError looks exactly like it would if it were generated by the server.

This patch mostly takes care of the linked issue. There'll be two more small patches to tie up loose ends once this merges:
1. We'll want to do something similar for InitPut requests, even though SQL no longer uses these.
2. We'll want to handle cases where the read half needs to be served out of the buffer.

Informs https://github.com/cockroachdb/cockroach/issues/139055

Release note: None

arulajmani committed:
kvclient: decompose CPut requests in the txnWriteBuffer

This patch decomposes CPut requests, which are read-write in nature,
into a read half and a write half. This is done by sending a (Exclusive)
locking Get request to KV, followed by the evaluation of the CPut
conditional on the client. If successful, the associated write is
buffered on the client. Subsequently, at flush time, this buffered
write will be converted to a Put request.

If the conditional evaluation isn't successful, an ConditionFailedError
is returned back. The layers above should be oblivious to this
transformation in the txnWriteBuffer; so the ConditionFailedError looks
exactly like it would if it were generated by the server.

This patch mostly takes care of the linked issue. There'll be two more
small patches to tie up loose ends once this merges:
1. We'll want to do something similar for InitPut requests, even though
SQL no longer uses these.
2. We'll want to handle cases where the read half needs to be served
out of the buffer.

Informs https://github.com/cockroachdb/cockroach/issues/139055

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142934)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! Just some minor comments from me.

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-:-OLVrSK3BHM3_kllpy-D:b-m712bc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @stevendanna)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVnuoy7Lw_FLlhiJZz:-OLVnuoy7Lw_FLlhiJ_-:bbp58jw):*
What's the rationale for supporting `InitPut`? We'll only enable buffered writes from SQL client when the cluster version has been finalized at 25.2 at which point we know there won't be any `InitPut`s issued ever.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVpNVS0cmT-vOtl0AS:-OLVpNVS0cmT-vOtl0AT:b-7xftxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8dca429a8bd3f4b1403613aa613330afd559c8a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L120)):*
> ```Go
> 
> 	// testingOverrideCPutEvalFn is used to mock the evaluation function for
> 	// conditional puts. Indended only for tests.
> ```

nit: `s/Indenden/intended/`.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVpbjf6FP7MYgZFQJg:-OLVpbjf6FP7MYgZFQJh:bjoq27a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8dca429a8bd3f4b1403613aa613330afd559c8a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L324)):*
> ```Go
> // difference is the direction in which the buffer is iterated when doing the
> // merge. As a result, they're also collected as tranformations.
> // 5. Conditional Puts are decomposed into a locking Get followed by a Put. The
> ```

nit: 3rd and 4th sentences seem to be almost a repeat of what's said in the 2nd sentence in this paragraph.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 814 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVqogL8202jg-hTwEO:-OLVqogL8202jg-hTwEP:bx90071) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6ca127585f892727cceed9bc65901226ac9908a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L814)):*
> ```Go
> 	// Assert that no buffered write was returned.
> 	require.Len(t, br.Responses[0].GetInner().(*kvpb.ScanResponse).Rows, 0)
> 	require.Len(t, br.Responses[1].GetInner().(*kvpb.ReverseScanResponse).Rows, 0)
> ```

nit: the removal looks like a merge skew.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVr1vG8QhGHkuKYcLZ:-OLVr1vG8QhGHkuKYcL_:b-i5h2zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8dca429a8bd3f4b1403613aa613330afd559c8a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L984)):*
> ```Go
> 		valA := "val"
> 
> 		// Blindly write to keys A.
> ```

nit: `s/keys/key/`.
___
*[`pkg/storage/mvcc.go` line 2990 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLVqEXl88cKsycxSGka:-OLVqEXl88cKsycxSGkb:b-ktwbvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8dca429a8bd3f4b1403613aa613330afd559c8a/pkg/storage/mvcc.go#L2990)):*
> ```Go
> 	expBytes []byte, actVal optionalValue, allowNoExisting bool,
> ) *kvpb.ConditionFailedError {
> 	return mvcceval.MaybeConditionFailedError(expBytes, actVal.ToPointer(), actVal.IsPresent(), allowNoExisting)
> ```

nit: should we inline this call in order to remove the helper method that doesn't seem useful anymore?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
This looks pretty good to me. I can handle Yahor's nit's and then I'll follow up with the locking get support on top of it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-:-OLZyoVSDaSQd62NlD4F:btjqah3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)



<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142934)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)

___
*[`pkg/storage/mvcceval/eval.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLi6DIX5YWpysrVW-p7:-OLi6DIX5YWpysrVW-p8:bsjzuya) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8dca429a8bd3f4b1403613aa613330afd559c8a/pkg/storage/mvcceval/eval.go#L19)):*
> ```Go
> // then a non-existent actual value is allowed even when expected-value is
> // non-empty.
> func MaybeConditionFailedError(
> ```

nit: Perhaps CheckExpectedValue


<!-- Sent from Reviewable.io -->

stevendanna committed:
kvclient: set LockNonExisting when necessary in txnWriteBuffer

When a ConditionalPut expects (or allows) no existing row, the initial
locking Get should lock non-existing rows.

Informs #139055

Release note: None

yuzefovich reviewed (approved):
Thanks!

> Reviewed 4 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-:-OLlBIn1-2cr_KJs01ej:b-ucfknh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 366 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142934#-OLlAETB-WAErsIx4vkz:-OLlAETB-WAErsIx4vl-:bamy3gq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d26562ece242da47b1c82816b2eff20a9523c1d/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L366)):*
> ```Go
> 			// TODO(arul): We're not handling the case where this Get needs to
> 			// be served locally from the buffer yet.
> ```

nit: let's keep the empty line for symmetry with other `case`s.


<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=yuzefovich
